pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/creator/JsonIndexCreator.java

```diff
@@ -29,7 +29,39 @@
   char KEY_VALUE_SEPARATOR = '\0';
 
   /**
-   * Adds the next json value.
+   * To be invoked once before each physical JSON document when streaming JSON into the index
+   * @param numFlattenedDocs the number of flattened documents to be streamed into the index
+   */
+  default void startDocument(int numFlattenedDocs) {
+  }
+
+  /**
+   * To be invoked after each physical JSON document when streaming JSON into the index.
+   */
+  default void endDocument() {
+  }
+
+  /**
+   * To be invoked once before each nested (logical) JSON document when streaming JSON into the index
```

Comment:
I am a bit confused here. What do you mean by `physical JSON document` and `nested (logical) JSON document`. I thought there was only one kind of JSON document (as defined by JSON spec)?

Reply:
By physical document I mean the actual JSON document being indexed. By nested (logical) JSON document I mean one of the flattened documents which actually gets indexed by the JSON index. Perhaps physical and logical aren't helping here.

Reply:
`flattened` isn't helping here either IMHO. Unless I am missing something there is no such thing as a `physical`, `nested`, or `flattened` JSON and use of such terms and interface functions associated with them will only cause confusion. We just need `start`, `end`, and `add` methods. Anything else is just very specific to your particular implementation of JSON index and should not be put at interface level and may not carry over to other implementations.

Reply:
Without understanding the concept of flattening it would be impossible to implement the existing JSON index, which is why it is modelled as a first class concept.

Reply:
I'm inclined to agree on this point and think I can achieve the same effect with a different interface. I will address this.

Reply:
I don't think this is resolved yet. Please give me some time to look at this further.